File: wzip/src/main/java/com/wwdablu/soumya/wzip/WUnzipWorker.java
package com.wwdablu.soumya.wzip;



import androidx.annotation.NonNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

class WUnzipWorker extends BaseWorker {

    WUnzipWorker(@NonNull File zipFile,
                 @NonNull File destinationFolder,
                 @NonNull String workerIdentifier,
                 @NonNull WZipCallback callback) {

        super(zipFile, destinationFolder, workerIdentifier, callback);
    }

    @Override
    public void run() {

        ZipInputStream zipInputStream = null;

        try {

            mCallback.onStarted(mWorkerIdentifier);
            createDestinationFolderIfMissing(mDestinationFolder);

            zipInputStream = new ZipInputStream(new FileInputStream(mZipFile));
            ZipEntry zipEntry = zipInputStream.getNextEntry();

            while (zipEntry != null) {

                FileOutputStream fos = new FileOutputStream(mDestinationFolder.getAbsolutePath() + File.separator + zipEntry.getName());
                int size = zipInputStream.read();

                while (size != -1) {
                    fos.write(size);
                    size = zipInputStream.read();
                }

                fos.close();
                zipEntry = zipInputStream.getNextEntry();
            }

            mCallback.onUnzipCompleted(mWorkerIdentifier);

        } catch (Throwable throwable) {
            mCallback.onError(throwable, mWorkerIdentifier);

        } finally {

            if(zipInputStream != null) {
                try {
                    zipInputStream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: wzip/src/main/java/com/wwdablu/soumya/wzip/WUnzipWorker.kt
package com.wwdablu.soumya.wzip

import android.content.Context
import androidx.documentfile.provider.DocumentFile
import java.util.zip.ZipInputStream

internal class WUnzipWorker(
    private val context: Context,
    private val zipFile: DocumentFile,
    private val destinationFolder: DocumentFile,
    private val workerIdentifier: String,
    private val callback: WZipCallback
) : BaseWorker() {
    override fun run() {

        context.contentResolver.openInputStream(zipFile.uri).use { zipFileInputStream ->

            ZipInputStream(zipFileInputStream).use { zis ->

                try {
                    callback.onStart(workerIdentifier, WZipCallback.Mode.UNZIP)

                    var zipEntry = zis.nextEntry
                    while (zipEntry != null) {

                        val opFile = destinationFolder.createFile("text/plain", zipEntry.name)
                        if (opFile != null) {
                            opFile.renameTo(zipEntry.name)
                            context.contentResolver.openOutputStream(opFile.uri).use { fos ->
                                var size = zis.read()
                                while (size != -1 && fos != null) {
                                    fos.write(size)
                                    size = zis.read()
                                }
                            }
                        }

                        zipEntry = zis.nextEntry
                    }

                    callback.onUnzipComplete(workerIdentifier, destinationFolder)

                } catch (exception: Exception) {
                    callback.onError(workerIdentifier, exception, WZipCallback.Mode.UNZIP)
                }
            }
        }
    }
}
